### PR TITLE
chore: enable exhaustruct linter for database param structs

### LINKED ISSUE #9936: Ensure all DB update struct values are explicit
Per https://github.com/coder/coder/issues/9823, it's possible to do an update like:

```go
type UpdateSomething struct {
   A string
   B string
}

func foo() {
   db.Update(UpdateSomething{
      A: "blah",
   })
}

```

and accidentally set `B` to its zero value. In #9823, this type of bug caused the loss of provisioner state, bricking workspaces—a worst-case scenario.

A simple resolution to this category of bug is adding an [exhauststruct](https://github.com/GaijinEntertainment/go-exhaustruct) check to our linting of form:

```
exhaustruct -i 'github.com/coder/coder/v2/coderd/database\..*Params' -test=false ./...
```

### PR DESCRIPTION
**Pull request by johnstcn**
Fixes https://github.com/coder/coder/issues/9936.

I chose to only enable this for database parameter structs that do not start with the string `Get`, as having to put in `//nolint` directives for all the times we specify `GetFooParams{}` would be exceedingly obnoxious.

### CODE REVIEW

File: coderd/audit.go [hunk 1]
```diff
@@ -154,6 +154,9 @@ func (api *API) generateFakeAuditLog(rw http.ResponseWriter, r *http.Request) {
 		Diff:             diff,
 		StatusCode:       http.StatusOK,
 		AdditionalFields: params.AdditionalFields,
+		OrganizationID:   uuid.Nil,
```
Comment (mtojek) [inline coderd/audit.go:157]: Oh, shouldn't we insert OrganizationID? I'm thinking about this: https://github.com/coder/coder/commit/9e845213f59f2eae704d15e1c69a0ad9a5dbe4b6

File: coderd/organizations.go [hunk 2]
```diff
@@ -68,10 +68,11 @@ func (api *API) postOrganizations(rw http.ResponseWriter, r *http.Request) {
...
 		organization, err = tx.InsertOrganization(ctx, database.InsertOrganizationParams{
-			ID:        uuid.New(),
-			Name:      req.Name,
-			CreatedAt: dbtime.Now(),
-			UpdatedAt: dbtime.Now(),
+			ID:          uuid.New(),
+			Name:        req.Name,
+			CreatedAt:   dbtime.Now(),
+			UpdatedAt:   dbtime.Now(),
+			Description: "",
```
Comment (mtojek) [inline coderd/organizations.go:75]: It looks like something we could expose via HTTP request.

File: coderd/users.go [hunk 3]
```diff
@@ -1079,10 +1079,11 @@ func (api *API) CreateUser(ctx context.Context, store database.Store, req Create
...
 			organization, err := tx.InsertOrganization(ctx, database.InsertOrganizationParams{
-				ID:        uuid.New(),
-				Name:      req.Username,
-				CreatedAt: dbtime.Now(),
-				UpdatedAt: dbtime.Now(),
+				ID:          uuid.New(),
+				Name:        req.Username,
+				CreatedAt:   dbtime.Now(),
+				UpdatedAt:   dbtime.Now(),
+				Description: "",
```
Comment (mtojek) [inline coderd/users.go:1086]: Same here

**Review by mtojek — COMMENTED** [review]
Left a couple of questions as you could discover another issues.

File: coderd/audit.go (hunk 1 above)
Comment (johnstcn) [reply coderd/audit.go:157]: Good point, makes sense to at least insert something fake here.

File: coderd/organizations.go (hunk 2 above)
Comment (mtojek) [reply coderd/organizations.go:75]: Maybe we should remove it? Of course, could be a separate PR.

Comment (johnstcn) [reply coderd/organizations.go:75]: Just the field, or the concept entirely? The field itself is probably just a forklift from v1. I feel that the entire concept also is, but that's a much wider topic.

Comment (mtojek) [reply coderd/organizations.go:75]: Yes, the field, but keep in mind that maybe I just don't see the full picture.

Comment (johnstcn) [reply coderd/organizations.go:75]: However, some [isuses](https://github.com/coder/coder/issues/8870) do mention exposing organizations. So maybe best to leave the field as-is for now.

**Review by mtojek — APPROVED** [review]
Thanks for addressing comments!

File: coderd/httpmw/apikey.go [hunk 4]
```diff
@@ -369,13 +369,15 @@ func ExtractAPIKey(rw http.ResponseWriter, r *http.Request, cfg ExtractAPIKeyCon
 		// If the API Key is associated with a user_link (e.g. Github/OIDC)
 		// then we want to update the relevant oauth fields.
 		if link.UserID != uuid.Nil {
-			// nolint:gocritic
+			// nolint:gocritic // system needs to update user link
```
Comment (mafredri) [inline coderd/httpmw/apikey.go:372]:
```suggestion
			//nolint:gocritic // system needs to update user link
```

This, I believe is the standard, but we have mixed use 😔 `//nolint` 289 vs `// nolint` 178.

File: coderd/provisionerdserver/provisionerdserver.go [hunk 5]
```diff
@@ -652,7 +655,12 @@ func (s *server) UpdateJob(ctx context.Context, request *proto.UpdateJobRequest)
 
 	if len(request.Logs) > 0 {
 		insertParams := database.InsertProvisionerJobLogsParams{
-			JobID: parsedID,
+			JobID:     parsedID,
+			CreatedAt: []time.Time{},
+			Source:    []database.LogSource{},
+			Level:     []database.LogLevel{},
+			Stage:     []string{},
+			Output:    []string{},
```
Comment (mafredri) [inline coderd/provisionerdserver/provisionerdserver.go:663]: Using `nil` is probably OK for these, considering they were missing before?

File: coderd/provisionerdserver/provisionerdserver.go [hunk 6]
```diff
@@ -1027,6 +1035,7 @@ func (s *server) CompleteJob(ctx context.Context, completed *proto.CompletedJob)
 		}
 
 		var completedError sql.NullString
+		var completedErrorCode sql.NullString
```
Comment (mafredri) [inline coderd/provisionerdserver/provisionerdserver.go:1038]: This doesn't seem to be ever assigned, it's use could be confusing for future change(r)s.

File: enterprise/replicasync/replicasync.go [hunk 7]
```diff
@@ -420,11 +420,13 @@ func (m *Manager) Close() error {
...
-		Hostname:     m.self.Hostname,
-		Version:      m.self.Version,
-		Error:        m.self.Error,
+		RelayAddress:    m.self.RelayAddress,
+		RegionID:        m.self.RegionID,
+		Hostname:        m.self.Hostname,
+		Version:         m.self.Version,
+		Error:           m.self.Error,
+		DatabaseLatency: m.self.DatabaseLatency,
+		Primary:         m.self.Primary,
```
Comment (mafredri) [inline enterprise/replicasync/replicasync.go:429]: Was this a bug, now fixed?

**Review by mafredri — APPROVED** [review]
Nice work, thanks for making this change!

File: enterprise/replicasync/replicasync.go (hunk 7 above)
Comment (johnstcn) [reply enterprise/replicasync/replicasync.go:429]: Now that I stare at this a little longer, this change itself may be a bug.
I don't see how it makes sense to have a non-zero database latency for a closed replica, or a closed replica be anything other than non-primary.

File: coderd/provisionerdserver/provisionerdserver.go (hunk 6 above)
Comment (johnstcn) [reply coderd/provisionerdserver/provisionerdserver.go:1038]: Removing it.

File: coderd/provisionerdserver/provisionerdserver.go [hunk 8]
```diff
@@ -271,10 +271,12 @@ func (s *server) AcquireJobWithCancel(stream proto.DRPCProvisionerDaemon_Acquire
 					Time:  dbtime.Now(),
 					Valid: true,
 				},
+				UpdatedAt: dbtime.Now(),
```
Comment (spikecurtis) [inline coderd/provisionerdserver/provisionerdserver.go:274]: nit: 2 calls to `Now()` might be separated by a few microseconds (or longer, if the goroutine is suspended in the middle), which might create an odd outcome where the job was updated later than it was completed

File: coderd/provisionerdserver/provisionerdserver.go [hunk 9]
```diff
@@ -652,7 +655,12 @@ func (s *server) UpdateJob(ctx context.Context, request *proto.UpdateJobRequest)
 
 	if len(request.Logs) > 0 {
 		insertParams := database.InsertProvisionerJobLogsParams{
-			JobID: parsedID,
+			JobID:     parsedID,
```
Comment (spikecurtis) [inline coderd/provisionerdserver/provisionerdserver.go:658]: nit: nolint is probably clearer in this case

**Review by spikecurtis — APPROVED** [review]

**Pull request merged**